### PR TITLE
Add `_block_refresh()` to layers

### PR DESCRIPTION
**Pull request by jamesyan-git**
# Description
As per [this discussion](https://github.com/napari/napari/pull/5861/files/ea2ac8c9364855cfff62a5692aab7f214ddd95bc#r1320937393) on the graph layer PR, this PR add a context manager to block `refresh()` from updating the view slice. 
We thought it would be best to separate this from the graph layer PR, so it isn't being snuck in within a larger PR.

Currently, this blocker is not used, however see [here](https://github.com/JoOkuma/napari/pull/31/files#diff-71346a695f04f54e1b48ad7224d72933f6bd6d99f46280b3024d83c48c7338d3R493) for how it will be used in a future PR.

Co-authored with @DragaDoncila

### CODE REVIEW

File: napari/layers/_tests/test_layer_attributes.py [hunk 1]
```diff
@@ -138,3 +140,40 @@ def test_get_value_3d_view_of_2d_image(ImageClass):
...
+        MyCounter.count += 1
+
+    my_layer.events.reload.connect(counter_increment)
+
+    with my_layer._block_refresh():
+        my_layer.refresh()
+    assert MyCounter.count == 0
+
+    my_layer.refresh()
+    assert MyCounter.count == 1
```
Comment (Czaki) [inline napari/layers/_tests/test_layer_attributes.py]:
```suggestion
    mock = MagicMock()

    my_layer.events.reload.connect(mock)

    with my_layer._block_refresh():
        my_layer.refresh()
        
    mock.assert_not_called()

    my_layer.refresh()
	mock.assert_called_once()
```

File: napari/layers/_tests/test_layer_attributes.py [hunk 2]
```diff
@@ -138,3 +140,40 @@ def test_get_value_3d_view_of_2d_image(ImageClass):
...
+
+    my_layer.refresh()
+    my_layer.set_view_slice.assert_called_once()
+
+
+@pytest.mark.parametrize('Layer, data, ndim', layer_test_data)
+def test_async_refresh_block(Layer, data, ndim):
+    from napari import settings
+
+    settings.get_settings().experimental.async_ = True
```
Comment (Czaki) [inline napari/layers/_tests/test_layer_attributes.py:162]: Why is this line needed?

Comment (jamesyan-git) [reply napari/layers/_tests/test_layer_attributes.py:162]: @Czaki This test checks that [this codepath](https://github.com/napari/napari/blob/main/napari/layers/base/base.py#L1409) is also being handled correctly by the blocker and it's only executed when experimental async is on. We (@DragaDoncila) also did find [the async fixture](https://github.com/napari/napari/blob/main/napari/_qt/_tests/test_async_slicing.py#L27) and initially moved it to `conftest` but it requests `make_napari_viewer` and so it broke headless tests, so we just decided to set async manually here. Let me know if we should approach this a different way.

Comment (Czaki) [reply napari/layers/_tests/test_layer_attributes.py:162]: But you put before this code. 

```python
        if self._refresh_blocked:
            logger.debug('Layer.refresh blocked: %s', self)
            return
```

Comment (DragaDoncila) [reply napari/layers/_tests/test_layer_attributes.py:162]: @Czaki I'm not entirely sure what your objection is - can you please expand? We added that `if` statement to implement the blocking, but wanted to write the tests to ensure both the sync refresh and async refresh were blocked. 

If you're concerned about editing the settings directly, I think the `fresh_settings` [fixture](https://github.com/napari/napari/blob/main/napari/conftest.py#L195) will take care of that. Otherwise though we could move the test out to a separate file that can make use of the `enable_async` fixture.

Comment (Czaki) [reply napari/layers/_tests/test_layer_attributes.py:162]: Ok. I see. This is a check for future refactor

**Review by Czaki — APPROVED** [review]

**Comment by DragaDoncila** [comment]
Will merge this now so we can update #4991. Thanks @Czaki
